contracts/truefi2/AaveBaseRateOracle.sol

```diff
@@ -0,0 +1,161 @@
+
+    IAaveLendingPool public aavePool;
+
+    address asset;
+
+    // A fixed amount of time to wait
+    // to be able to update the totalsBuffer
+    uint256 public cooldownTime;
+
+    /**
+     * @dev Throws if cooldown is on when updating the totalsBuffer
+     */
+    modifier offCooldown() {
+        // get the last timestamp written into the buffer
+        uint256 lastWritten = totalsBuffer.timestamps[totalsBuffer.latestIndex];
+        require(block.timestamp >= lastWritten.add(cooldownTime), "AaveBaseRateOracle: Buffer on cooldown");
+        _;
+    }
+
+    constructor(
+        IAaveLendingPool _aavePool,
+        uint256 _cooldownTime,
+        address _asset
+    ) public {
+        aavePool = _aavePool;
+        cooldownTime = _cooldownTime;
+        asset = _asset;
+
+        totalsBuffer.lastValue = getAaveVariableBorrowAPY(asset);
+        totalsBuffer.timestamps[0] = block.timestamp;
+    }
+
+    function bufferSize() public virtual pure returns (uint16) {
+        return MAX_BUFFER_SIZE;
+    }
+
+    /**
+     * @dev Helper function to get contents of the totalsBuffer
+     */
+    function getTotalsBuffer()
+        public
+        view
+        returns (
+            uint256[MAX_BUFFER_SIZE] memory,
+            uint256[MAX_BUFFER_SIZE] memory,
+            uint16
+        )
+    {
+        return (totalsBuffer.runningTotals, totalsBuffer.timestamps, totalsBuffer.latestIndex);
+    }
+
+    function getAaveVariableBorrowAPY(address _asset) public view returns (uint256) {
+        (, , , , uint128 currentVariableBorrowRate, , , , , , , ) = aavePool.getReserveData(_asset);
+        return uint256(currentVariableBorrowRate).div(1e23);
+    }
+
+    /**
+     * @dev Update the totalsBuffer:
+     * Gets current deposit apy from aave and writes down
+     * new total running value.
+     * If the buffer is filled overwrites the oldest value
+     * with a new one and updates its timestamp.
+     */
+    function update() public offCooldown {
+        uint16 lidx = totalsBuffer.latestIndex;
+        uint16 nextIndex = (lidx + 1) % bufferSize();
+        uint256 apy = getAaveVariableBorrowAPY(asset);
+        uint256 dt = block.timestamp.sub(totalsBuffer.timestamps[lidx]);
+        totalsBuffer.runningTotals[nextIndex] = totalsBuffer.runningTotals[lidx].add(apy.add(totalsBuffer.lastValue).mul(dt).div(2));
+        totalsBuffer.timestamps[nextIndex] = block.timestamp;
+        totalsBuffer.lastValue = apy;
+        totalsBuffer.latestIndex = nextIndex;
+    }
+
+    /**
+     * @dev Average apy is calculated by taking
+     * the time-weighted average of the aave deposit apys.
+     * Essentially formula given below is used:
+     *
+     *           (v + v_{n-1}) / 2 * (t - t_{n-1}) + sum_{i=1}^{n - 1} (v_i + v_{i-1}) / 2 * (t_i - t_{i-1})
+     * avgAPY = -------------------------------------------------------------------------------------------
+     *                                                  (t - t_0)
```

Comment:
LRS: Update comment to sum_{i=1}^n v_i * (t_i - t_{i-1}) on top and t_n - t_0 below.